src/main.rs

```diff
@@ -1,3 +1,131 @@
+
+    let challenge_name: Name = find_record(cli.identifier)?;
+    info!("challenge name is {}", challenge_name);
+
+    let zone: Name = find_zone(challenge_name.clone())?;
+    info!("challenge zone is {}", zone);
+
+    let primary_ns_client = &config
+        .zones
+        .get(&zone)
+        .context("Couldn't find challenge zone in config")?
+        .create_client();
+
+    let challenge_record = Record::from_rdata(
+        challenge_name.clone(),
+        Duration::minutes(1).whole_seconds() as u32,
+        RData::TXT(TXT::new(Vec::from([cli.proof]))),
+    );
+
+    match &cli.command.unwrap() {
+        cli::Commands::Set => {
+            let result = primary_ns_client
+                .create(challenge_record.clone(), zone)
+                .context("Couldn't set record")?;
+            match result.response_code() {
+                ResponseCode::NoError => {
+                    info!("record was successfully set");
+                    Ok(())
+                }
+                ResponseCode::YXRRSet => {
+                    warn!("{:?} already exists", challenge_record.clone());
+                    Ok(())
+                }
+                _ => todo!(),
+            }
+        }
+        cli::Commands::Cleanup => {
+            let result = primary_ns_client
+                .delete_rrset(challenge_record.clone(), zone)
+                .context("Couldn't remove record")?;
+            match result.response_code() {
+                ResponseCode::NoError => {
+                    info!("record was removed");
+                    Ok(())
+                }
+                _ => {
+                    debug!("{:?}", result);
+                    Ok(())
+                }
+            }
+        }
+    }
+}
+
+fn find_record(identifier: Name) -> anyhow::Result<Name> {
+    let prefix: Name = Name::from_str("_acme-challenge").unwrap();
+    let q = happy::search(
+        prefix.append_name(&identifier)?, // this should be called prepend because that's what it actualyl does
+        RecordType::TXT,
+    )?;
+    let last_record = q.answers().last().unwrap().clone().into_parts();
+    let challenge_name = match last_record.rdata.unwrap() {
+        RData::CNAME(name) => name.0,
+        RData::TXT(_) => {
+            info!(
+                "found existing TXT record for {:?}",
+                last_record.name_labels
+            );
+            last_record.name_labels
+        }
+        _ => panic!("unexpected record response"),
+    };
+    Ok(challenge_name)
+}
+
+fn find_zone(identifier: Name) -> anyhow::Result<Name> {
+    fn from_name_labels(response: DnsResponse) -> Option<Name> {
+        response
+            .name_servers()
+            .iter()
+            .map(|record| record.clone().into_parts().name_labels)
+            .last()
+    }
+
+    let q = happy::search(identifier.clone(), RecordType::SOA)?;
+    let name = match q.response_code() {
+        ResponseCode::NoError => from_name_labels(q),
```

Comment:
Does that always work? Because if it does, then we shouldn't need the match statement at all...

Reply:
The function that gets called here can return an empty result, in which case it gets caught later. I don't think this was a good way of implementing it seeing it's difficult to read.